File: haskell/lib/Lens_combinators.hs
import Data.List (intercalate)

data RE = Str String | Re String deriving (Show, Eq)

data Lens a
  = Del RE String
  | Store RE
  | Value String
  | Counter String
  | Seq String
  | Key RE
  | Label String
  deriving (Show, Eq)

concat :: Lens a -> Lens a -> [String] -> [String]
concat l1 l2 = concatMap (l2 . concat l1 l2)

union :: Lens a -> Lens a -> [String] -> [String]
union l1 l2 = (++) <$> l1 <*> l2

repetition :: String -> Lens a -> [String] -> [String]
repetition op l = case op of
  "*" -> concatMap (++ repetition op l) . l
  "+" -> concatMap (++ repetition op l) . filter (not . null) . l
  "?" -> (++ repetition op l) <$> l
  _ -> error "Invalid repetition operator"

subtree :: Lens a -> [String] -> [[String]]
subtree l = filter (not . null) . map l

square :: Lens a -> Lens a -> Lens a -> [String] -> [String]
square left body right = concatMap process
  where
    process s =
      case left s of
        [] -> []
        ls ->
          case right ls of
            [] -> []
            rs -> body (drop (length ls) $ init rs)

File: ocaml/lib/lens_combinators.ml
type 're = Str of string | Re of string

type 'a lens =
  | Del of 're * string
  | Store of 're
  | Value of string
  | Counter of string
  | Seq of string
  | Key of 're
  | Label of string

let rec concat l1 l2 = function
  | [] -> []
  | s :: rest ->
     (match l1 s with
      | [] -> concat l2 (s :: rest)
      | l1_res ->
         match concat l2 l1_res with
         | [] -> l1_res
         | l2_res -> l2_res @ concat (concat l1 l2) rest)

let union l1 l2 s =
  match l1 s with
  | [] -> l2 s
  | res -> res

let rec repetition op l = function
  | [] -> []
  | s :: rest ->
     (match op with
      | '*' -> (match l s with
                | [] -> repetition op l rest
                | res -> res @ repetition op l rest)
      | '+' -> (match l s with
                | [] -> []
                | res -> res @ repetition op l rest)
      | '?' -> (match l s with
                | [] -> repetition op l rest
                | res -> res @ repetition op l rest)
      | _ -> invalid_arg "Invalid repetition operator")

let subtree l =
  let rec process_subtree = function
    | [] -> []
    | s :: rest ->
       (match l s with
        | [] -> process_subtree rest
        | res -> res :: process_subtree rest)
  in
  process_subtree

let square left body right s =
  let rec process_square = function
    | [] -> []
    | s :: rest ->
       (match left s with
        | [] -> process_square rest
        | left_res ->
           match right left_res with
           | [] -> process_square rest
           | right_res ->
              let mid = List.rev left_res @ right_res in
              (body mid) @ process_square rest)
  in
  process_square s